src/natcap/invest/habitat_quality.py

```diff
@@ -551,264 +620,337 @@ def execute(args):
+    pygeoprocessing.reclassify_raster(
+        new_cover_path, code_index, rarity_path, gdal.GDT_Float32,
+        _OUT_NODATA)
 
-            # a dictionary to map LULC types to a number that depicts how
-            # rare they are considered
-            code_index = {}
+    LOGGER.info(
+        'Finished rarity computation on %s land cover.',
+        os.path.basename(lulc_path_band[0]))
 
-            # compute rarity index for each lulc code
-            # define 0.0 if an lulc code is found in the cur/fut landcover
-            # but not the baseline
-            for code in lulc_code_count_x:
-                if code in lulc_code_count_b:
-                    numerator = lulc_code_count_x[code] * lulc_area
-                    denominator = lulc_code_count_b[code] * base_area
-                    ratio = 1.0 - (numerator / denominator)
-                    code_index[code] = ratio
-                else:
-                    code_index[code] = 0.0
 
-            rarity_path = os.path.join(
-                out_dir, 'rarity' + lulc_key + suffix + '.tif')
+def _create_decay_kernel(raster_path_band, kernel_path, decay_type, max_dist):
+    """Create a decay kernel as a raster.
 
-            pygeoprocessing.reclassify_raster(
-                (new_cover_path, 1), code_index, rarity_path, gdal.GDT_Float32,
-                _RARITY_NODATA)
+    Args:
+        raster_path_band (tuple): a 2 tuple of the form
+            (filepath to raster, band index) for raster to decay.
+        kernel_path (string): path to output kernel raster.
+        decay_type (string): type of decay kernel to create, either
+            'linear' | 'exponentional'.
+        max_dist (float): max distance of threat in KM.
 
-            LOGGER.info('Finished rarity computation on %s land cover.'
-                        % lulc_time)
-    LOGGER.info('Finished habitat_quality biophysical calculations')
+    Returns:
+        None
+    """
+    # need the pixel size for the raster so we can create an appropriate
+    # kernel for convolution
+    threat_pixel_size = pygeoprocessing.get_raster_info(
+        raster_path_band[0])['pixel_size']
+
+    # convert max distance (given in KM) to meters
+    max_dist_m = max_dist * 1000.0
+
+    # convert max distance from meters to the number of pixels that
+    # represents on the raster
+    max_dist_pixel = max_dist_m / abs(threat_pixel_size[0])
+    LOGGER.debug('Max distance in pixels: %f', max_dist_pixel)
+
+    # blur the raster based on the decay type
+    if decay_type == 'linear':
+        decay_func = _make_linear_decay_kernel_path
+    elif decay_type == 'exponential':
+        decay_func = utils.exponential_decay_kernel_raster
+    else:
+        raise ValueError(
+            "Unknown type of decay in biophysical table, should be"
+            f" either 'linear' or 'exponential'. Input was {decay_type}"
+            " for threat"
+            f" os.path.splitext(os.path.basename(raster_path_band[0]))[0]))"
```

Comment:
I think the content of this f-string needs to be wrapped in `{ }` or else that's a fun error message to get. :P

Reply:
Fixed here: https://github.com/natcap/invest/pull/75/commits/9413243a540a9b62e840cefb0749c2e680e617fd